### PR TITLE
fix: building in release mode for simulator

### LINKED ISSUE #29984: After updating to Xcode 12 I get The linked library 'libPods-*****.a' is missing one or more architectures required by this target: arm64.
## Description
After updating to Xcode 12 I get this error:

 `The linked library 'libPods-*****.a' is missing one or more architectures required by this target: arm64.`

Anyone knows how can solve it?

### PR DESCRIPTION
**Pull request by grabbou**
## Summary

Fixes #29984

Right now, running a React Native application with Xcode 12 in Release mode on an iPhone Simulator will fail with something like below:

> [some file path], building for iOS Simulator, but linking in object file built for iOS, file '[some file path]' for architecture arm64

The best explanation of this issue has been provided by @alloy in #29984:

> This issue has started coming up with Xcode 12 and support for the new ARM based Macs, as `arm64` now no longer can be assumed to _only_ be for iOS devices. This means Xcode 12 will now also build for `arm64` simulator SDKs and it has become ambiguous if an arch slice in a prebuilt binary is meant for a simulator or device.
> 
> In any case, for now this means that you can configure your Xcode project to exclude `arm64` when building for any iOS simulator SDK.

This PR implements aforementioned workaround.

## Changelog

[FIX] [IOS] - Fix running React Native project with Xcode 12 in Release on iPhone Simulator

## Test Plan

Switch your scheme to Release and run the app on simulator. Will complete w/o issues.

### CODE REVIEW

**Review by alloy — APPROVED** [review]
Yup 👍

**Comment by andreialecu** [comment]
I'm on an M1 MacBook Mini and if I add `arm64` to excluded architectures I'm getting this error right at the beginning of the build:

```
Module map file '/Users/.../Library/Developer/Xcode/DerivedData/MyApp-bmibskzjfhlxbsblspkqkqtndyjb/Build/Products/Debug-iphonesimulator/YogaKit/YogaKit.modulemap' not found
```

It seems that running the build through Rosetta via `arch -x86_64 react-native run-ios` does work though, but there's a glitch with momentum scrolling, which seems like a bug when arm64 slices are missing according to this report:
https://developer.apple.com/forums/thread/668488

It also feels relatively slow, as if it runs on the simulator via rosetta x86 simulation rather than being native ARM.

So, how do we build for the simulator and keep the `arm64` arch?

**Comment by andreialecu** [comment]
The error in the previous comment seems to go away if I change the deployment target to `11.0` from `10.0` (as per: https://github.com/facebook/react-native/issues/29605#issuecomment-701920944) but then I got this which is related to Flipper:

```
In /Users/andreialecu/Work/.../ios/Pods/OpenSSL-Universal/ios/lib/libcrypto.a(cryptlib.o), building for iOS Simulator, but linking in object file built for iOS, file '/Users/andreialecu/Work/.../ios/Pods/OpenSSL-Universal/ios/lib/libcrypto.a' for architecture arm64
```

Seems like I needed to manually edit the `Pods` project to add `arm64` to its excluded architectures as well and then it works. 

However, the glitch with momentum scrolling remains because of the missing `arm64` slice. And the app in the simulator is pretty janky because of it.

**Comment by andreialecu** [comment]
Also note that I'm not even attempting to run in `Release` mode, just `Debug` -> but I felt the PR was relevant.

**Comment by andreialecu** [comment]
I was able to get an `arm64` build working on the M1 Mac that works in the iOS simulator by adding this in the `Podfile`.

Also needed to disable Flipper.

```ruby
  # Enables Flipper.
  #
  # Note that if you have use_frameworks! enabled, Flipper will not work and
  # you should disable these next few lines.

  #use_flipper!
  post_install do |installer|
    #flipper_post_install(installer)

    installer.pods_project.targets.each do |target|
      target.build_configurations.each do |config|
        config.build_settings.delete 'ARCHS'
        config.build_settings.delete 'IPHONEOS_DEPLOYMENT_TARGET'
        config.build_settings['ONLY_ACTIVE_ARCH'] = 'YES'
      end
    end
  end
```

I'm not sure if this also works on non-M1 Macs, but it would be something to try.

**Comment by andreialecu** [comment]
Upon further experimentation it seems the `Podfile` changes above may not be necessary **except** for disabling Flipper. I was not able to get Flipper to work.

One other change I needed to do was to remove `LIBRARY_SEARCH_PATHS` from my `project.pbxproj`:

```diff
@@ -1007,11 +1017,6 @@
                                GCC_WARN_UNUSED_VARIABLE = YES;
                                IPHONEOS_DEPLOYMENT_TARGET = 10.0;
                                LD_RUNPATH_SEARCH_PATHS = "/usr/lib/swift $(inherited)";
-                               LIBRARY_SEARCH_PATHS = (
-                                       "\"$(TOOLCHAIN_DIR)/usr/lib/swift/$(PLATFORM_NAME)\"",
-                                       "\"$(TOOLCHAIN_DIR)/usr/lib/swift-5.0/$(PLATFORM_NAME)\"",
-                                       "\"$(inherited)\"",
-                               );
                                MTL_ENABLE_DEBUG_INFO = YES;
                                ONLY_ACTIVE_ARCH = YES;
                                SDKROOT = iphoneos;
@@ -1060,11 +1065,6 @@
                                GCC_WARN_UNUSED_VARIABLE = YES;
                                IPHONEOS_DEPLOYMENT_TARGET = 10.0;
                                LD_RUNPATH_SEARCH_PATHS = "/usr/lib/swift $(inherited)";
-                               LIBRARY_SEARCH_PATHS = (
-                                       "\"$(TOOLCHAIN_DIR)/usr/lib/swift/$(PLATFORM_NAME)\"",
-                                       "\"$(TOOLCHAIN_DIR)/usr/lib/swift-5.0/$(PLATFORM_NAME)\"",
-                                       "\"$(inherited)\"",
-                               );
                                MTL_ENABLE_DEBUG_INFO = NO;
                                SDKROOT = iphoneos;
                                VALIDATE_PRODUCT = YES;
```
I'm not sure if the above was part of the RN default template or not, but with it I was running into weird problems with Swift Pods.

Also needed to manually add a Swift Bridging Header to the main project.

Now the app properly runs on the M1 simulator in ARM mode and it's responsive and fast.

The main takeaway is that I think you should really consider whether disabling `arm64` is the proper solution to this problem, because it may have unintended consequences on the M1 macs.

**Comment by alloy** [comment]
@andreialecu Thanks for all the feedback! To be clear, this is absolutely it meant as the long term solution, but as a workaround to still being able to release 0.64 without further delay and serve the common case out of the box (those without ARM based macs). Longer term the solution is to switch to producing xcframeworks artefacts instead.

**Comment by andreialecu** [comment]
@alloy I noticed this was recently merged: https://github.com/facebook/flipper/pull/1735 and it seems like a fix for the same problem. Unfortunately I don't have an Intel Mac any more to test it.

**Comment by alloy** [comment]
Yup, just saw that too. I’ll see how I can best test/verify that.

File: scripts/react_native_pods.rb [hunk 1]
```diff
@@ -108,3 +108,24 @@ def flipper_post_install(installer)
...
+    .map{ |t| t.user_project }
+    .uniq{ |p| p.path }
+    .push(installer.pods_project)
+
+  arm_value = `/usr/sbin/sysctl -n hw.optional.arm64 2>&1`.to_i
+
+  projects.each do |project|
+    project.build_configurations.each do |config|
+      if arm_value == 1 then
+        config.build_settings["EXCLUDED_ARCHS[sdk=iphonesimulator*]"] = "arm64"
```
Comment (andreialecu) [inline scripts/react_native_pods.rb:123]: I might be reading this wrong somehow, but isn't this `if` the wrong way around? 

Seems to exclude `arm64` when on `arm`.

**Comment by andreialecu** [comment]
I was able to update Flipper to the one containing the fix via:

```rb
use_flipper!('Flipper-Folly' => { :podspec => 'https://raw.githubusercontent.com/facebook/flipper/master/iOS/Podspecs/Flipper-Folly.podspec' } )
```

```
Analyzing dependencies
Fetching podspec for `Flipper-Folly` from `https://raw.githubusercontent.com/facebook/flipper/master/iOS/Podspecs/Flipper-Folly.podspec`
Downloading dependencies
Installing Flipper-Folly 2.3.0
Installing OpenSSL-Universal 1.1.171 (was 1.0.2.20)
Generating Pods project
Integrating client project
```


However, linking on the M1 still fails, but now with a different dependency:

```
in .../ios/Pods/CocoaLibEvent/lib/libevent.a(buffer.o), building for iOS Simulator, but linking in object file built for iOS, file '.../ios/Pods/CocoaLibEvent/lib/libevent.a' for architecture arm64
```

`CocoaLibEvent` seems to also be a dependency of `Flipper`: 

https://github.com/facebook/flipper/blob/031cec299b527975cd1e252177372bb33cb0ae18/iOS/Podspecs/Flipper-Folly.podspec#L20
https://github.com/e314521/CocoaLibEvent

I'm not very familiar with pods or xcode in general so I'm not sure why the lib is incompatible or how it can be fixed.

The Flipper PR thus seems incomplete.

File: scripts/react_native_pods.rb (hunk 1 above)
Comment (grabbou) [reply scripts/react_native_pods.rb:123]: Oh, true!

**Comment by grabbou** [comment]
Basically, Flipper fix was to exclude `arm64` as we are doing in this PR. As a result, the error started happening in the next library after it.

I am going to double check the `if` statement and let's ship this PR for broader testing.

**Pull request closed without merging**